File: backend/src/main/java/com/solo/video/service/TagService.java
package com.solo.video.service;

import com.solo.video.dto.request.TagCreateRequest;
import com.solo.video.dto.response.TagResponse;

import java.util.List;

public interface TagService {
    
    List<TagResponse> getAllTags(String keyword, String sortBy, String sortOrder);
    
    TagResponse getTagById(Long id);
    
    TagResponse createTag(TagCreateRequest request);
    
    TagResponse updateTag(Long id, TagCreateRequest request);
    
    void deleteTag(Long id);
    
    void addTagsToVideo(Long videoId, List<Long> tagIds);
    
    void removeTagFromVideo(Long videoId, Long tagId);
    
    void addTagsToVideos(List<Long> videoIds, List<Long> tagIds);
}

File: frontend/src/services/tagService.ts
import api from './api'
import { Tag, ApiResponse, TagCreateRequest } from '@/types'

export const tagService = {
  getTags: async (params: {
    keyword?: string
    sortBy?: string
    sortOrder?: string
  } = {}): Promise<Tag[]> => {
    const response = await api.get<ApiResponse<Tag[]>>('/tags', { params })
    return response.data.data
  },

  createTag: async (request: TagCreateRequest): Promise<Tag> => {
    const response = await api.post<ApiResponse<Tag>>('/tags', request)
    return response.data.data
  },

  updateTag: async (id: number, request: Partial<TagCreateRequest>): Promise<Tag> => {
    const response = await api.put<ApiResponse<Tag>>(`/tags/${id}`, request)
    return response.data.data
  },

  deleteTag: async (id: number): Promise<void> => {
    await api.delete(`/tags/${id}`)
  },

  addTagsToVideo: async (videoId: number, tagIds: number[]): Promise<void> => {
    await api.post(`/tags/video/${videoId}`, { tagIds })
  },

  removeTagFromVideo: async (videoId: number, tagId: number): Promise<void> => {
    await api.delete(`/tags/video/${videoId}/${tagId}`)
  },

  addTagsToVideos: async (videoIds: number[], tagIds: number[]): Promise<void> => {
    await api.post('/tags/videos/batch', { videoIds, tagIds })
  },
}
